File: core/src/main/java/com/netflix/iceberg/FilteredManifest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.netflix.iceberg;

import com.google.common.base.Preconditions;
import com.google.common.collect.Iterables;
import com.google.common.collect.Iterators;
import com.netflix.iceberg.ManifestEntry.Status;
import com.netflix.iceberg.expressions.Evaluator;
import com.netflix.iceberg.expressions.Expression;
import com.netflix.iceberg.expressions.Expressions;
import com.netflix.iceberg.expressions.InclusiveMetricsEvaluator;
import com.netflix.iceberg.expressions.Projections;
import java.util.Collection;
import java.util.Iterator;

public class FilteredManifest implements Filterable<FilteredManifest> {
  private final ManifestReader reader;
  private final Expression partFilter;
  private final Expression rowFilter;
  private final Collection<String> columns;

  // lazy state
  private Evaluator lazyEvaluator = null;
  private InclusiveMetricsEvaluator lazyMetricsEvaluator = null;

  FilteredManifest(ManifestReader reader, Expression partFilter, Expression rowFilter,
                   Collection<String> columns) {
    Preconditions.checkNotNull(reader, "ManifestReader cannot be null");
    this.reader = reader;
    this.partFilter = partFilter;
    this.rowFilter = rowFilter;
    this.columns = columns;
  }

  @Override
  public FilteredManifest select(Collection<String> columns) {
    return new FilteredManifest(reader, partFilter, rowFilter, columns);
  }

  @Override
  public FilteredManifest filterPartitions(Expression expr) {
    return new FilteredManifest(reader,
        Expressions.and(partFilter, expr),
        rowFilter,
        columns);
  }

  @Override
  public FilteredManifest filterRows(Expression expr) {
    Expression projected = Projections.inclusive(reader.spec()).project(expr);
    return new FilteredManifest(reader,
        Expressions.and(partFilter, projected),
        Expressions.and(rowFilter, expr),
        columns);
  }

  Iterable<ManifestEntry> allEntries() {
    if ((rowFilter != null && rowFilter != Expressions.alwaysTrue()) ||
        (partFilter != null && partFilter != Expressions.alwaysTrue())) {
      Evaluator evaluator = evaluator();
      InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();

      return Iterables.filter(reader.entries(columns),
          entry -> (entry != null &&
              evaluator.eval(entry.file().partition()) &&
              metricsEvaluator.eval(entry.file())));

    } else {
      return reader.entries(columns);
    }
  }

  Iterable<ManifestEntry> liveEntries() {
    if ((rowFilter != null && rowFilter != Expressions.alwaysTrue()) ||
        (partFilter != null && partFilter != Expressions.alwaysTrue())) {
      Evaluator evaluator = evaluator();
      InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();

      return Iterables.filter(reader.entries(columns),
          entry -> (entry != null &&
              entry.status() != Status.DELETED &&
              evaluator.eval(entry.file().partition()) &&
              metricsEvaluator.eval(entry.file())));

    } else {
      return Iterables.filter(reader.entries(columns),
          entry -> entry != null && entry.status() != Status.DELETED);
    }
  }

  @Override
  public Iterator<DataFile> iterator() {
    if ((rowFilter != null && rowFilter != Expressions.alwaysTrue()) ||
        (partFilter != null && partFilter != Expressions.alwaysTrue())) {
      Evaluator evaluator = evaluator();
      InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();

      return Iterators.transform(
          Iterators.filter(reader.iterator(partFilter, columns),
              input -> (input != null &&
                  evaluator.eval(input.partition()) &&
                  metricsEvaluator.eval(input))),
          DataFile::copy);

    } else {
      return Iterators.transform(reader.iterator(partFilter, columns), DataFile::copy);
    }
  }

  private Evaluator evaluator() {
    if (lazyEvaluator == null) {
      if (partFilter != null) {
        this.lazyEvaluator = new Evaluator(reader.spec().partitionType(), partFilter);
      } else {
        this.lazyEvaluator = new Evaluator(reader.spec().partitionType(), Expressions.alwaysTrue());
      }
    }
    return lazyEvaluator;
  }

  private InclusiveMetricsEvaluator metricsEvaluator() {
    if (lazyMetricsEvaluator == null) {
      if (rowFilter != null) {
        this.lazyMetricsEvaluator = new InclusiveMetricsEvaluator(
            reader.spec().schema(), rowFilter);
      } else {
        this.lazyMetricsEvaluator = new InclusiveMetricsEvaluator(
            reader.spec().schema(), Expressions.alwaysTrue());
      }
    }
    return lazyMetricsEvaluator;
  }
}

File: python/iceberg/core/filtered_manifest.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from iceberg.api.expressions import Evaluator, Expressions, inclusive, InclusiveMetricsEvaluator

from .manifest_entry import Status


class FilteredManifest(object):

    def __init__(self, reader, part_filter, row_filter, columns, case_sensitive=True):
        if reader is None:
            raise RuntimeError("ManifestReader cannot be null")

        self.reader = reader
        self.part_filter = part_filter
        self.row_filter = row_filter
        self.columns = columns
        self.case_sensitive = case_sensitive

        self.lazy_evaluator = None
        self.lazy_metrics_evaluator = None

    def select(self, columns):
        return FilteredManifest(self.reader, self.part_filter, self.row_filter, columns, self.case_sensitive)

    def filter_partitions(self, expr):
        return FilteredManifest(self.reader,
                                Expressions.and_(self.part_filter, expr),
                                self.row_filter,
                                self.columns,
                                self.case_sensitive)

    def filter_rows(self, expr):
        projected = inclusive(self.reader.spec).project(expr)
        return FilteredManifest(self.reader,
                                Expressions.and_(self.part_filter, projected),
                                Expressions.and_(self.row_filter, expr),
                                self.columns, self.case_sensitive)

    def all_entries(self):
        if self.row_filter is not None and self.row_filter != Expressions.always_true() \
                or self.part_filter is not None and self. part_filter != Expressions.always_true():
            evaluator = self.evaluator()
            metrics_evaluator = self.metrics_evaluator()
            return [entry for entry in self.reader.entries(self.columns)
                    if entry is not None
                    and evaluator.eval(entry.file.partition())
                    and metrics_evaluator.eval(entry.file)]
        else:
            return self.reader.entries(self.columns)

    def live_entries(self):
        if self.row_filter is not None and self.row_filter != Expressions.always_true() \
                or self.part_filter is not None and self. part_filter != Expressions.always_true():
            evaluator = self.evaluator()
            metrics_evaluator = self.metrics_evaluator()
            return [entry for entry in self.reader.entries(self.columns)
                    if entry is not None
                    and entry.status != Status.DELETED
                    and evaluator.eval(entry.file.partition())
                    and metrics_evaluator.eval(entry.file)]
        else:

            return [entry for entry in self.reader.entries(self.columns)
                    if entry is not None and entry.status != Status.DELETED]

    def iterator(self):
        if self.row_filter is not None and self.row_filter != Expressions.always_true() \
                or self.part_filter is not None and self.part_filter != Expressions.always_true():
            evaluator = self.evaluator()
            metrics_evaluator = self.metrics_evaluator()

            return (input.copy() for input in self.reader.iterator(self.part_filter, self.columns)
                    if input is not None
                    and evaluator.eval(input.partition())
                    and metrics_evaluator.eval(input))
        else:
            return (entry.copy() for entry in self.reader.iterator(self.part_filter, self.columns))

    def evaluator(self):
        if self.lazy_evaluator is None:
            if self.part_filter is not None:
                self.lazy_evaluator = Evaluator(self.reader.spec.partition_type(),
                                                self.part_filter,
                                                self.case_sensitive)
            else:
                self.lazy_evaluator = Evaluator(self.reader.spec.partition_type(),
                                                Expressions.always_true(),
                                                self.case_sensitive)

        return self.lazy_evaluator

    def metrics_evaluator(self):
        if self.lazy_metrics_evaluator is None:
            if self.row_filter is not None:
                self.lazy_metrics_evaluator = InclusiveMetricsEvaluator(self.reader.spec.schema,
                                                                        self.row_filter, self.case_sensitive)
            else:
                self.lazy_metrics_evaluator = InclusiveMetricsEvaluator(self.reader.spec.schema,
                                                                        Expressions.always_true(),
                                                                        self.case_sensitive)

        return self.lazy_metrics_evaluator
